internal/configs/provider_validation.go

```diff
@@ -831,3 +842,165 @@ func providerName(name, alias string) string {
 	}
 	return name
 }
+
+func providerIterationIdenticalWarning(blockType string, sourceExpr, instanceExpr hcl.Expression) hcl.Diagnostics {
+	var diags hcl.Diagnostics
+	if providerIterationIdentical(sourceExpr, instanceExpr) {
+		// foot, meet gun
+		diags = append(diags, &hcl.Diagnostic{
+			Severity: hcl.DiagWarning,
+			Summary:  "Likely misconfiguration of provider iteration",
+			Detail:   fmt.Sprintf("Provider and %s both share identical iteration expressions, this is not recommended. When a key is removed, the corresponding resources will no longer have a provider that is able to destroy them.\nInstead, it is recommended to have %s's iteration be a subset of the provider's iteration expression. See the OpenTofu documentation for more details.\n\nTo disable this warning, wrap one of the expressions in a function like coalesce() to disable the check.", blockType, blockType),
+			Subject:  sourceExpr.Range().Ptr(),
+		})
+	}
```

Comment:
Just some tweaking to try to make this sound more like the existing "OpenTofu diagnostics voice":

```suggestion
		diags = append(diags, &hcl.Diagnostic{
			Severity: hcl.DiagWarning,
			Summary:  "Provider configuration for_each matches " + blockType, 
			Detail:   fmt.Sprintf(
				"This provider configuration uses the same for_each expression as %s, which means that subsequent removal of elements from this collection would cause a planning error.\n\nOpenTofu relies on a provider instance to destroy resource instances that are associated with it, and so the provider instance must outlive all of its resource instances by at least one plan/apply round. For removal of instances to succeed in future you must structure the configuration so that the provider block's for_each expression can produce a superset of the instances of the resources associated with the provider configuration. Refer to the OpenTofu documentation for specific suggestions.",
				referrerAddr,
			),
			Subject:  sourceExpr.Range().Ptr(),
		})
	}
```

(note that this message wants to include the full address of the thing whose for_each we were comparing against, rather than just the block type name, so the signature of this function would need to change to include that. I'm proposing that it start with something like "The provider configuration uses the same for_each expression as aws_instance.example..." to be explicit about what we were comparing with.)

---

I was also thinking about ways we could potentially make this warning less annoying without requiring authors to write confusing patterns like a single-argument `coalesce` to silence the warning.

One potential answer would be to limit this warning only to two situations:
- When we've produced a plan that includes at least one "create" action for an affected resource.
- When we're returning the "Provider instance not present" error during provider resolution, as an extra hint as to the likely cause of the error.

Of course, that means we'd need to deal with this check at runtime in `package tofu`, rather than in `package configs`. I think covering both of the situations above would require calling into this check (and thus potentially generating the warning) in two places:

- In the error handling codepath of provider resolution, at the same time we're adding this diagnostic:

     https://github.com/opentofu/opentofu/blob/7645992d9b44d42e2767f4e34ffafded8b172556/internal/tofu/node_resource_abstract_instance.go#L213-L220

- As part of the series of special checks we perform once we've determined the "planned action" for a particular resource instance, nearby this one:

    https://github.com/opentofu/opentofu/blob/7645992d9b44d42e2767f4e34ffafded8b172556/internal/tofu/node_resource_abstract_instance.go#L1228-L1239

The main advantage of implementing it this way is that someone can intentionally ignore the warning and apply anyway, in which case the warning will remain quiet until it becomes relevant again rather than bothering the operator on every run until someone introduces a confusing workaround in the configuration.

A secondary advantage is that we'd always be reporting the warning in terms of a specific single managed resource and never for a whole module, which means that the warning would not appear if calling a module that, for example, consists only of `data` blocks fetching data elsewhere. (There is no problem in that case, because data resource instances are never "destroyed" in the sense that managed resource instances are.)

Reply:
My preference is to leave this where it is currently. I'd rather warn the user as early as possible that they have likely done something wrong before they get into trouble.  You do make a good point about data blocks, I should exclude them from this check.

Reply:
If we intend to always return this warning on every run, even when the affected objects aren't changing, then I think we ought to offer a more explicit way to quiet the warning that isn't an abuse of a function that has some other meaning. Folks often inherit OpenTofu modules written by other people who are no longer around to explain themselves, and so quirky workarounds tend to become maintenance hazards.

I don't think that using the `coalesce` function qualifies because a `coalesce` call with only one argument would, to anyone who is relying only on [the documentation for that function](https://opentofu.org/docs/language/functions/coalesce/), seem like a mistake rather than something intentional. :thinking: 

However, I suppose we could launch without such a thing and see if anyone requests a way to quiet the warning. If our heuristic is good enough that it never produces false-positives then this would be a moot point.